recipes/zarp/meta.yaml

```diff
@@ -0,0 +1,70 @@
+{% set name = "zarp" %}
+{% set version = "0.1.1" %}
+
+package:
+  name: "{{ name|lower }}"
+  version: "{{ version }}"
+
+source:
+  url: "https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz"
+  sha256: 0defdec6fe8dabf66d40f2f4d16dfc2619f509f11763356908af3be51ca049ab
+
+build:
+  noarch: python
+  entry_points:
+   - zarp = zarp.cli:main
+  number: 0
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  host:
+    - addict >=2.4.0
+    - bidict >=0.22.0
+    - email-validator >=1.2.1
+    - jsonref >=0.2
+    - pandas >=1.3.5
+    - pydantic >=1.9.2 
+    - pygments >=2.8.0
+    - python
+    - rich >=12.5.1
+    - pip
+    - pip:
+      - pydantic-collections >=0.3.0
+  run:
+    - addict >=2.4.0
+    - bidict >=0.22.0
+    - email-validator >=1.2.1
+    - jsonref >=0.2
+    - pandas >=1.3.5
+    - pydantic >=1.9.2
+    - pygments >=2.8.0
+    - python
+    - rich >=12.5.1
+    - pip
+    - pip:
+      - pydantic-collections >=0.3.0
```

Comment:
`pip` makes sense to me in host, but I'm not sure if all the dependencies need to show up in both sections? Maybe the maintainers can help here?

Reply:
pip should go only to host, and pip-only dependencies are not allowed.
They should be turned into conda packages before. In this case a conda-forge package. I can help with this if you ping me over at conda-forge.

Reply:
Thanks @bgruening, appreciated! @fgypas has already raised a PR: https://github.com/conda-forge/staged-recipes/pull/22285#issuecomment-1468889508

Linting is failing atm, @fgypas will look again tomorrow. Will ping you if we find it difficult to bug.

To clarify: apart from `pip` (`host` only), will we need all the dependencies in both `host` and `run`?

Reply:
The lint it failing over there. And no you don't need pip in the run requirements.

Reply:
Sorry if I wasn't being clear. Guess my question was: Will we need _all_ of the `run` requirements _also_ in `host`? Or _just_ `pip` and maybe `python`?

But after re-reading https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#requirements-section, I tend to think the answer is that we need to put everything that we need to build in `host` (or `build`), and everything to run in `run`. And I think that we _do_ need all the dependencies to build, so I guess the answer is "yes, we do need to specify most dependencies twice"

Reply:
@uniqueg I usually look at other packages and try to get inspiration. For example, https://github.com/bioconda/bioconda-recipes/blob/master/recipes/bakta/meta.yaml is a good one. No need to reinvent anything here.

Our documentation also recommends using the skeleton generator and I think this tool is available on pypi, so you can simply use the `conda skeleton pypi` and `grayskull pypi` ...

Let me know if you need anything else.

Reply:
Thanks @bgruening.

`conda skeleton pypi` is precisely what produced these `host` and `run` requirements. The example you put, on the other hand, seems to _not_ repeat the `run` requirements in `host`. But of course it's a different software, so it may be fine like that there and not here. OR `conda skeleton pypi` is unable to understand what is or isn't required in `host` and just plays it safe by just fetching all of the requirements and PyPI and putting them in both sections. It's hard to say, and I didn't find clarification in the docs on this one (or at least I didn't understand enough so that I'd feel confident to make a call here), hence my question. Not trying to reinvent anything, just wanting to avoid unnecessary repitition and, if possible, increase my understanding of `meta.yaml`.

Reply:
This is what the skeleton generator spits out for me:

```
{% set name = "zarp" %}
{% set version = "0.1.1" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/zarp-{{ version }}.tar.gz
  sha256: 0defdec6fe8dabf66d40f2f4d16dfc2619f509f11763356908af3be51ca049ab

build:
  entry_points:
    - zarp = zarp.cli:main
  noarch: python
  script: {{ PYTHON }} -m pip install . -vv
  number: 0

requirements:
  host:
    - python
    - setuptools_git ==1.2
    - pip
  run:
    - python
    - addict ==2.4.0
    - bidict ==0.22.0
    - email-validator ==1.2.1
    - jsonref ==0.2
    - pandas ==1.3.5
    - pydantic ==1.9.2
    - pydantic-collections ==0.3.0
    - pygments ==2.8.0
    - rich ==12.5.1

test:
  imports:
    - zarp
  commands:
    - pip check
    - zarp --help
  requires:
    - pip

about:
  home: https://git.scicore.unibas.ch/zavolan_group/tools/zarp-cli
  summary: User-friendly command-line interface for the ZARP RNA-Seq analysis pipeline
  license: Apache-2.0
  license_file: LICENSE

extra:
  recipe-maintainers:
    - bgruening
 ```
 
 These pins do look too strict to me, but this is what upstream is defining. Is the pip package working in general?

Reply:
Hmm, I got this (with `conda` 22.11.1 and `conda-build` 3.23.3, both from `conda-forge`): 

```yaml
{% set name = "zarp" %}
{% set version = "0.1.1" %}

package:
  name: "{{ name|lower }}"
  version: "{{ version }}"

source:
  url: "https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz"
  sha256: 0defdec6fe8dabf66d40f2f4d16dfc2619f509f11763356908af3be51ca049ab

build:
  number: 0
  noarch: python
  entry_points:
    - zarp = zarp.cli:main
  script: "{{ PYTHON }} -m pip install . -vv"

requirements:
  host:
    - addict ==2.4.0
    - bidict ==0.22.0
    - email-validator ==1.2.1
    - jsonref ==0.2
    - pandas ==1.3.5
    - pip
    - pydantic ==1.9.2
    - pydantic-collections ==0.3.0
    - pygments ==2.8.0
    - python
    - rich ==12.5.1
  run:
    - addict ==2.4.0
    - bidict ==0.22.0
    - email-validator ==1.2.1
    - jsonref ==0.2
    - pandas ==1.3.5
    - pydantic ==1.9.2
    - pydantic-collections ==0.3.0
    - pygments ==2.8.0
    - python
    - rich ==12.5.1

test:
  imports:
    - tests
    - tests.config
    - zarp
    - zarp.config
    - zarp.extensions
    - zarp.run
  commands:
    - zarp --help

about:
  home: "https://git.scicore.unibas.ch/zavolan_group/tools/zarp-cli"
  license: Apache Software
  license_family: APACHE
  license_file: 
  summary: "User-friendly command-line interface for the ZARP RNA-Seq analysis pipeline"
  doc_url: 
  dev_url: 

extra:
  recipe-maintainers:
    - your-github-id-here
```

Yours looks much better :)

Yeah, the pinned versions are from upstream, and you are right in that they shouldn't be pinned that strictly. I'll prepare a new release tomorrow and we can take it from there.

Reply:
And yes, pip is working. But will double check tomorrow, loosen the pins, fix a few other things, create a new release and then will update the branch here. Thanks for your help!